x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -0,0 +1,173 @@
+func (s *StorageManager) RunGCLoop(stopping <-chan struct{}, gcInterval time.Duration) error {
+	select {
+	case <-stopping:
+		return nil
+	case s.gcLoopCh <- struct{}{}:
+	}
+	defer func() {
+		<-s.gcLoopCh
+	}()
+	// This goroutine is responsible for periodically garbage
+	// collecting the Badger value log, using the recommended
+	// discard ratio of 0.5.
+	ticker := time.NewTicker(gcInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-stopping:
+			return nil
+		case <-ticker.C:
+			const discardRatio = 0.5
+			var err error
+			for err == nil {
+				// Keep garbage collecting until there are no more rewrites,
+				// or garbage collection fails.
+				err = s.runValueLogGC(discardRatio)
+			}
+			if err != nil && err != badger.ErrNoRewrite {
+				return err
+			}
+		}
+	}
+}
+
+func (s *StorageManager) runValueLogGC(discardRatio float64) error {
+	return s.db.RunValueLogGC(discardRatio)
+}
+
+func (s *StorageManager) Close() error {
+	s.rw.Close()
+	return s.db.Close()
+}
+
+// Size returns the db size
+func (s *StorageManager) Size() (lsm, vlog int64) {
+	return s.db.Size()
+}
+
+func (s *StorageManager) ReadSubscriberPosition() ([]byte, error) {
+	s.subscriberPosMu.Lock()
+	defer s.subscriberPosMu.Unlock()
+	return os.ReadFile(filepath.Join(s.storageDir, subscriberPositionFile))
+}
+
+func (s *StorageManager) WriteSubscriberPosition(data []byte) error {
+	s.subscriberPosMu.Lock()
+	defer s.subscriberPosMu.Unlock()
+	return os.WriteFile(filepath.Join(s.storageDir, subscriberPositionFile), data, 0644)
+}
+
+func (s *StorageManager) NewReadWriter() *ManagedReadWriter {
+	return &ManagedReadWriter{
+		sm: s,
+	}
+}
+
+// ManagedReadWriter is a read writer that is transparent to badger DB changes done by StorageManager.
+// It is a wrapper of the ShardedReadWriter under StorageManager.
+type ManagedReadWriter struct {
+	sm *StorageManager
+}
+
+func (s *ManagedReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error {
+	return s.sm.rw.ReadTraceEvents(traceID, out)
+}
+
+func (s *ManagedReadWriter) WriteTraceEvent(traceID, id string, event *modelpb.APMEvent, opts WriterOpts) error {
+	return s.sm.rw.WriteTraceEvent(traceID, id, event, opts)
+}
+
+func (s *ManagedReadWriter) WriteTraceSampled(traceID string, sampled bool, opts WriterOpts) error {
+	return s.sm.rw.WriteTraceSampled(traceID, sampled, opts)
+}
+
+func (s *ManagedReadWriter) IsTraceSampled(traceID string) (bool, error) {
+	return s.sm.rw.IsTraceSampled(traceID)
+}
+
+func (s *ManagedReadWriter) DeleteTraceEvent(traceID, id string) error {
+	return s.sm.rw.DeleteTraceEvent(traceID, id)
+}
+
+func (s *ManagedReadWriter) Flush() error {
+	return s.sm.rw.Flush()
+}
+
+// NewBypassReadWriter returns a ReadWriter directly reading and writing to the database,
+// bypassing any wrapper e.g. ShardedReadWriter.
+// This should be used for testing only, useful to check if data is actually persisted to the DB.
+func (s *StorageManager) NewBypassReadWriter() *ReadWriter {
```

Comment:
@1pkg I thought about it again. Merging rw interface into StorageManager makes it slightly more confusing especially now I introduce a BypassReadWriter for testing purposes. It is clearer that you can create a default shared read writer using NewReadWriter (which wraps ShardedReadWriter -> lockedReadWriter -> ReadWriter), and alternatively a NewBypassReadWriter.